source/ast/expressions/AssertionExpr.cpp

```diff
@@ -594,26 +718,93 @@ AssertionExpr& SequenceConcatExpr::fromSyntax(const DelayedSequenceExprSyntax& s
     return *result;
 }
 
-bool SequenceConcatExpr::admitsEmptyImpl() const {
-    auto it = elements.begin();
-    SLANG_ASSERT(it != elements.end());
+bitmask<NondegeneracyStatus> SequenceConcatExpr::checkNondegeneracyImpl() const {
+    bool admitsEmpty = true;
+    bool admitsNoMatch = false;
+    auto left = elements.begin();
+    SLANG_ASSERT(left != elements.end());
+    auto leftNondegenSt = left->sequence->checkNondegeneracy();
 
-    // See F.3.4.2.2 for the rules here.
-    if (it->delay.min != 0 || !it->sequence->admitsEmpty())
-        return false;
+    if (left->delay.min != 0 || !leftNondegenSt.has(NondegeneracyStatus::AdmitsEmpty))
+        admitsEmpty = false;
 
-    while (++it != elements.end()) {
-        if (!it->sequence->admitsEmpty())
-            return false;
+    if (leftNondegenSt.has(NondegeneracyStatus::AdmitsNoMatch))
+        admitsNoMatch = true;
 
-        if (it->delay.min == 0u && it->delay.max == 0u)
-            return false;
+    // See SystemVerilog LRM sections 16.9.2.1 and F.3.4.2.2 for details
+    while (admitsEmpty || !admitsNoMatch) {
```

Comment:
Does this condition need to be `&&` instead of `||` ? I'm wondering if you can see a sequence that makes you set admitsEmpty to false and then you end the loop, whereas if you had kept going you would have found that a later sequence makes it actually accept no match at all?

Reply:
Currently I haven't fix that because concat sequences can be no match without empty subsequences for example:

```verilog
sequence s;
    1'b0 ##1 seq1 ##2 seq2  // empty because of first `false` don't give an any chance to match it - `1'b0`
endsequence
```
If we change loop condition from `||` to `&&` we may have a very limited info about certain type of sequences. For example:

```verilog
sequence s;
    1'b0 ##1 seq1 ##2 seq2[*0]  // this seq will be marked only as "no match" but not as "admits an empty matches"
endsequence
```